game/ffmpeg.hh

```diff
@@ -110,51 +33,110 @@ extern "C" {
 /// ffmpeg class
 class FFmpeg {
   public:
-	/// Decode file; if no rate is specified, decode video, otherwise decode audio.
-	FFmpeg(fs::path const& file, unsigned int rate = 0);
-	~FFmpeg();
-	void operator()(); ///< Thread runs here, don't call directly
-	unsigned width = 0; ///< width of video
-	unsigned height = 0; ///< height of video
-	/// queue for video
-	VideoFifo  videoQueue;
-	/// queue for audio
-	AudioBuffer  audioQueue;
-	/** Seek to the chosen time. Will block until the seek is done, if wait is true. **/
-	void seek(double time, bool wait = true);
+	// Exceptions thrown by class
+	class Eof: public std::exception {};
+	class Error;
+	friend Error;
+
+	/// Decode file, depending on media type audio.
+	FFmpeg(fs::path const& filename, int mediaType);
+
+	void handleOneFrame();
+
+	/** Seek to the chosen time. **/
+	virtual void seek(double time);
+
 	/// duration
 	double duration() const;
-	bool terminating() { return m_quit_future.wait_for(0s) == std::future_status::ready; }
 
-	class eof_error: public std::exception {};
-  private:
-	void seek_internal();
-	void open();
-        struct Packet;
-	void decodePacket(Packet &);
+	virtual ~FFmpeg() = default;
+        
+  protected:
 	static void frameDeleter(AVFrame *f) { if (f) av_frame_free(&f); }
 	using uFrame = std::unique_ptr<AVFrame, std::integral_constant<decltype(&frameDeleter), &frameDeleter>>;
-	void processVideo(uFrame frame);
-	void processAudio(uFrame frame);
+
+	virtual void processFrame(uFrame frame) = 0;
+
+	struct Packet;
+	void decodePacket(Packet &);
+        
 	static void avformat_close_input(AVFormatContext *fctx);
 	static void avcodec_free_context(AVCodecContext *avctx);
 
 	fs::path m_filename;
-	unsigned int m_rate = 0;
-	std::promise<void> m_quit;
-	std::future<void> m_quit_future = m_quit.get_future();
-	std::atomic<double> m_seekTarget{ getNaN() };
 	double m_position = 0.0;
 	double m_duration = 0.0;
 	// libav-specific variables
 	int m_streamId = -1;
-	int m_mediaType;  // enum AVMediaType
 	std::unique_ptr<AVFormatContext, decltype(&avformat_close_input)> m_formatContext{nullptr, avformat_close_input};
 	std::unique_ptr<AVCodecContext, decltype(&avcodec_free_context)> m_codecContext{nullptr, avcodec_free_context};
+};
+
+class AudioFFmpeg : public FFmpeg {
+  public:
+	using AudioCb = std::function<void(const std::int16_t *data, size_t count, int64_t sample_position)>;
```

Comment:
Why `int64_t`?
I think that is probably the reason why the macOS build is failing due to conflicting parameter types in std::min.

Reply:
For the 'signed' reason: position can be negative as start position in ffmpeg may not be 0 and thus the first frame decoded have negative timestamps.
For the size, well, 64 is 'large enough', may have been 32bits

Reply:
Then I guess you should static_cast count to int64_t before comparing them?

Reply:
I forced the type of `write_pos_in_ring` to `size_t` (see the patch added) so that the std::min is done on size_t and the compiler does not complain anymore.

Reply:
Fair enough.